### PR TITLE
Fix: Keep proxies in sync across "Select networks for each site" toggling on/off

### PR DESCRIPTION
**Pull request by adonesky1**
## Explanation

This PR fixes a bug where the proxies used to manage per dapp selected network connections do not stay in sync when the user toggles on or off the "Select networks for each site" preference (which turns request queueing on and off):
Screenshot 2024-04-12 at 10 54 52 AM

This PR fixes this bug by subscribing to `PreferenceController` changes and, when the "Select networks for each site" toggle is flipped repointing the proxy from the (nested) globally selected network proxy to its own independently managed proxy, and when flipped off repoint the proxy to the globally selected network proxy.

To test: pull into [this extension PR](https://github.com/MetaMask/metamask-extension/compare/ad/fix/queue-toggle-proxies?expand=1) -> you'll need to link to a local build of selected-network-controller with these commits.

### `@metamask/selected-network-controller`

- **ADD**:  **BREAKING:** A parameter `useRequestQueuePreference` which should point to the current preferences state for useRequestQueue is now required by the constructor
- **ADD**: An `onPreferencesStateChange` argument that should subscribe to `PreferencesController` state changes and call a callback with the updated state is now a required parameter in the constructor options object.
- **REMOVED** - The `getUseRequestQueue` parameter is no longer accepted by the constructor.

## Checklist

- [ ] I've updated the test suite for new or updated code as appropriate
- [ ] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [ ] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

**Comment by adonesky1** [comment]
@metamaskbot publish-preview

**Comment by adonesky1** [comment]
@metamaskbot publish-preview

File: packages/selected-network-controller/src/SelectedNetworkController.ts [hunk 1]
```diff
@@ -201,6 +204,33 @@ export class SelectedNetworkController extends BaseController<
...
+
+    onPreferencesStateChange(({ useRequestQueue }) => {
+      if (this.#useRequestQueuePreference !== useRequestQueue) {
+        if (!useRequestQueue) {
+          Object.keys(this.state.domains).forEach((domain) => {
+            this.#unsetNetworkClientIdForDomain(domain);
+          });
+        } else {
+          // eslint-disable-next-line
+          (this.#domainProxyMap as any).map.forEach(
```
Comment (adonesky1) [inline packages/selected-network-controller/src/SelectedNetworkController.ts:216]: 🤔 not sure the best way to make this not awkward and more agnostic to this custom class

Comment (adonesky1) [reply packages/selected-network-controller/src/SelectedNetworkController.ts:216]: added [a pass through forEach on the WeakRefObjectMap class](https://github.com/MetaMask/metamask-extension/pull/23525/files#diff-4d13f1f0ba4d949f6dfe026c7c50a982dfd0e3b43d3bbd265f6acdca8f501302R57) for now

File: packages/selected-network-controller/tests/SelectedNetworkController.test.ts [hunk 2]
```diff
@@ -593,4 +617,158 @@ describe('SelectedNetworkController', () => {
...
+          expect(messenger.call).toHaveBeenCalledWith(
+            'NetworkController:getNetworkClientById',
+            'mainnet',
+          );
+          expect(mockProxyProvider.setTarget).toHaveBeenCalledTimes(2);
+          expect(mockProxyBlockTracker.setTarget).toHaveBeenCalledTimes(2);
+        });
+      });
+      describe('when domains do not have permissions', () => {
+        it('does not set the target of the existing proxy to the networkClient for the globally selectedNetworkClientId', () => {
```
Comment (jiexi) [inline packages/selected-network-controller/tests/SelectedNetworkController.test.ts:690]:
```suggestion
        it('does not change the target of the existing proxy', () => {
```

Comment (jiexi) [reply packages/selected-network-controller/tests/SelectedNetworkController.test.ts:690]: or something else is fine. I think the `to the networkClient for the globally selectedNetworkClientId` bit is not accurate?

File: packages/selected-network-controller/tests/SelectedNetworkController.test.ts [hunk 3]
```diff
@@ -593,4 +617,158 @@ describe('SelectedNetworkController', () => {
...
+
+          triggerPreferencesStateChange({ useRequestQueue: true });
+
+          expect(mockProxyProvider.setTarget).toHaveBeenCalledTimes(0);
+          expect(mockProxyBlockTracker.setTarget).toHaveBeenCalledTimes(0);
+        });
+      });
+    });
+    describe('when toggled from on to off', () => {
+      it('loops through domains in the domainsProxyMap and sets the target of the existing proxies to the (itself proxied) networkClient for the globally selectedNetworkClient', () => {
```
Comment (jiexi) [inline packages/selected-network-controller/tests/SelectedNetworkController.test.ts:728]:
```suggestion
      it('sets the target of the existing proxies to the (itself proxied) globally selected networkClient', () => {
```

or something like this?

File: packages/selected-network-controller/tests/SelectedNetworkController.test.ts [hunk 4]
```diff
@@ -593,4 +617,158 @@ describe('SelectedNetworkController', () => {
...
+    const mockProxyBlockTracker = {
+      setTarget: jest.fn(),
+    } as unknown as BlockTrackerProxy;
+
+    afterEach(() => {
+      jest.clearAllMocks();
+    });
+    describe('when toggled from off to on', () => {
+      describe('when domains have permissions', () => {
+        it('loops through domains in the domainsProxyMap and sets the target of the existing proxies to an independent networkClient for the globally selected networkClientId', () => {
```
Comment (jiexi) [inline packages/selected-network-controller/tests/SelectedNetworkController.test.ts:640]:
```suggestion
        it('sets the target of the existing proxies to the non-proxied networkClient for the globally selected networkClientId', () => {
```

File: packages/selected-network-controller/src/SelectedNetworkController.ts [hunk 5]
```diff
@@ -201,6 +204,32 @@ export class SelectedNetworkController extends BaseController<
...
+            // can't use public setNetworkClientIdForDomain because it will throw an error
+            // rather than simply skip if the domain doesn't have permissions which can happen
+            // in this case since proxies are added for each site the user visits
+            if (this.#domainHasPermissions(domain)) {
+              this.#setNetworkClientIdForDomain(
+                domain,
+                selectedNetworkClientId,
+              );
+            }
+          });
```
Comment (jiexi) [inline packages/selected-network-controller/src/SelectedNetworkController.ts:228]: thoughts on moving this block of logic into a private method?

Comment (adonesky1) [reply packages/selected-network-controller/src/SelectedNetworkController.ts:228]: We could. Happy to do it. Don't feel strongly about it either way!

Comment (adonesky1) [reply packages/selected-network-controller/src/SelectedNetworkController.ts]: done here: https://github.com/MetaMask/core/pull/4130/commits/37fb0106594ba5664ce3ad33c8d2879caa5bbf32

**Review by jiexi — APPROVED** [review]

**Pull request merged**